Track pointer position and add Release Buttons keyword

hbatagelo opened the pull request:
Fixes #97 and #98.

New keywords:
* `Walk Pointer To ${destination}`
* `Walk Pointer To Proportional ${end_position}`
* `Get Pointer Position` (used by the keywords above)
* `Release Buttons`

Remark: when `WaylandHid` is created, the pointer is moved to (0,0) and the tracking starts from that position.

Saviq commented:
I'd say we should drop the `From…` variants, they just muddy the waters and may cause unexpected things.

I mean:

```
Move Pointer To ${template}
Walk Pointer To ${template_two}
```
vs.
```
Walk Pointer From ${template} To ${template_two}
```

There's an implicit `Move` at the beginning of the `Walk`, and there's not really that much less to type?

Saviq commented:
(toggled draft so it picks up the changes in `main`)

Saviq commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -21,6 +21,7 @@ class WaylandHid(VirtualPointer):
 
     def __init__(self) -> None:
         self.ROBOT_LIBRARY_LISTENER = self
+        self.pointer_position = {"x": 0.0, "y": 0.0}
```
Why do we deal with floats for x/y, Wayland [only does](https://wayland.app/protocols/wlr-virtual-pointer-unstable-v1#zwlr_virtual_pointer_v1:request:motion_absolute) int?

Also, since this is internal, let's prefix with `_`? On one hand a tuple could do, but then we're returning it as a dict often enough that it's probably appropriate to be a dict.

Saviq commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -29,12 +30,15 @@ async def move_pointer_to_absolute(self, x: float, y: float) -> None:
         """Move the virtual pointer to an absolute position within the output."""
         await self.connect()
         self.move_to_absolute(x, y)
+        self.pointer_position = {"x": x, "y": y}
```
Let's update in place?

```suggestion
        self.pointer_position |= {"x": x, "y": y}
```

hbatagelo replied:
TIL

Saviq commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -29,12 +30,15 @@ async def move_pointer_to_absolute(self, x: float, y: float) -> None:
     async def move_pointer_to_proportional(self, x: float, y: float) -> None:
         """Move the virtual pointer to a position proportional to the size of the output."""
         await self.connect()
         self.move_to_proportional(x, y)
+        absolute_position = await self.get_absolute_from_proportional(x, y)
+        self.pointer_position = absolute_position
```
Let's just return a tuple from `move_to_proportional`?

```suggestion
        self.pointer_position["x"], self.pointer_position["y"] = self.move_to_proportional(x, y)
        return self.pointer_position
```

Saviq commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -48,20 +52,33 @@ async def release_pointer_button(self, button: str) -> None:
         """Connect to the display."""
         if not self.pointer:
             await super().connect()
+            await self.move_pointer_to_absolute(self.pointer_position["x"], self.pointer_position["y"])
```
Shouldn't we actually move to `(0, 0)` and only set the current position here?

I mean, the position is undefined until now.

hbatagelo replied:
I think `pointer_position` is defined as `(0,0)` in `__init__`.

Saviq commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -48,20 +52,33 @@ async def release_pointer_button(self, button: str) -> None:
-        await self.connect()
         return {"x": x * self.output_width, "y": y * self.output_height}
 
+    @keyword
+    async def get_pointer_position(self) -> Dict[str, float]:
+        """Get the current pointer position."""
+        return self.pointer_position
```
This doesn't need to be async?

Saviq reviewed (changes requested):
Some comments inline.

Overall, I wonder if we should push that much into Robot keywords, rather than directly into the library?

I mean, a lot of the keywords in the `.resource` are implementation details, I would say we should limit the API surface to things we actually want exposed to the procedure.

Like, `Get Pointer Position`  wouldn't even be needed if `Walk…` was implemented in the library. WDYT?

hbatagelo commented:
I agree. I'll try to push Walk into the library, remove the clutter and hide the implementation details... and use tuples of ints instead of dictionaries.

hbatagelo commented:
Note to myself:
* `pointer_position` should start as `None` and we shouldn't move the pointer upon connection
* `Walk To` without a previous move should be an error

Saviq commented:
> I'll try to push Walk into the library

Realized there's at least one reason why some of those keywords actually fit into a resource… and that's combining functionality from multiple libraries (like `Move To ${template}`, `Walk To ${template}`). Don't think we necessarily want `WaylandHid` to hard-depend on `Screencopy`, but rather on its interface. Say we want to introduce a VNC client to read the frames.

I suppose it could be done through library arguments somehow, but not sure it's worth it. And any case, not this PR.

hbatagelo commented:
Latest changes (addresses code review feedback and fixes #102):

* The keywords `Move From ... To ...`, `Walk From ... To ...`, `Distance From...`, `Lerp...`, and `Get Pointer Position` have been removed.
* The pointer won't move upon connection anymore. It is now an error to try to walk without moving at least once.
* Documentation has been improved.

Now, in `WaylanHid`, we have:
* `Move Pointer To (${x}, ${y})`
* `Move Pointer To Proportional (${x}, ${y})`
* `Press ${button} Button`
* `Release ${button} Button`
* `Release Buttons`
Also available in the library, but wrapped by higher-level keywords in the resource file:
* `Walk Pointer To Absolute    ${x}    ${y}    ${steps}    ${delay}`
* `Walk Pointer To Proportional    ${x}    ${y}    ${steps}    ${delay}`

In the resource file:
* `Get Center of ${region}`
* `Displace ${point} By (${x}, ${y})`
* `Move Pointer To ${destination}`, where `${destination}` can be a template or Python tuple
* `Walk Pointer To ${destination}`, where `${destination}` can be a template or Python tuple
* `Walk Pointer To (${x}, ${y})`
* `Walk Pointer To Proportional (${x}, ${y})`

These last two had to stay in the resource file (as opposed to `Move Pointer To (${x}, ${y})` and `Move Pointer To Proportional (${x}, ${y})`) because of the optional arguments. The optional arguments are positional arguments, but unfortunately Robot doesn't allow mixing positional args with embedded args in the library.

---

Here's a small script for showing the different keyword syntaxes:
```
Move Pointer To (200, 100)
Move Pointer To ${SRC_TEMPLATE}
Move Pointer To Proportional (1, 1)
Walk Pointer To (200, 100)
Walk Pointer To ${SRC_TEMPLATE}
Walk Pointer To Proportional (1, 1)

${p1}=    Walk Pointer To Proportional (0.5, 0.5)
${p2}=    Displace ${p1} By (200, 50)
Walk Pointer To ${p2}            # Same as Walk Pointer To (${p2}[0], ${p2}[1])
Move Pointer To ${p1}            # Same as Move Pointer To (${p1}[0], ${p1}[1])
```

---

Some thoughts/remarks:

- Maybe all keywords with embedded arguments in `WaylandHid` should be moved to the resource file? By doing so, I think the user wouldn't need to refer to the interface of `WaylandHid` for most use cases.
- Walks with `delay=0` are faster in the library and fails many tests when run locally. Using `delay=0.01` fixed it.
- Unlike `Walk Pointer To`, `Walk Pointer To Proportional` doesn't currently support a Python tuple as input. However, implementing an overload for it is straightforward if necessary,

hbatagelo commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,100 @@ Library             ../robot_libraries/WaylandHid.py
 
-Move Pointer To ${destination}
-    [Documentation]    Move the pointer to a position or image template.
+Move Pointer To ${destination:^\(?!\(.+,.+\)\).+}
```
It's not pretty and maybe there's a better of doing it: it matches "anything that is not (x, y)" to distinguish it from `Move Pointer To (${x}, ${y})`. Alternatively, we could parse the string in Python, but it would be more work.

Saviq commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -19,44 +20,90 @@ class WaylandHid(VirtualPointer):
+            walk_y = (1 - t) * self._pointer_position[1] + t * y
+            self.move_to_absolute(walk_x, walk_y)
+            await asyncio.sleep(delay)
+
+        self._pointer_position = (x, y)
```
This is handled by `move_to_absolute` already.
```suggestion
```

hbatagelo replied:
Not really. `move_to_absolute` is from `VirtualPointer`, not the robot library.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    Arguments:
+    ...    - ${x}: Integer absolute x-coordinate to move the pointer to.
+    ...    - ${y}: Integer absolute y-coordinate to move the pointer to.
+    WaylandHid.Move Pointer To Absolute    ${x}    ${y}
```
If we introduce another Hid library, this would be easier:

```suggestion
    Move_Pointer_To_Absolute    ${x}    ${y}
```

hbatagelo replied:
Some of these will conflict with keywords with the same name in the resource file. Instead of dropping `WaylandHid`, I changed it to `HID` and imported the library `AS    HID`:
```robot
Library             ../robot_libraries/WaylandHid.py    AS    HID
```
WDYT?

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
         ${position}=    Set Variable    ${destination}
     END
-    WaylandHid.Move Pointer To Absolute    ${position}[x]    ${position}[y]
+    WaylandHid.Move Pointer To Absolute    ${position}[0]    ${position}[1]
```
```suggestion
    Move_Pointer_To_Absolute    ${position}[0]    ${position}[1]
```

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
-    ...    ${position} must be a point represented by floating-point x and y
-    ...    values in the range 0..1.
+    ...    Arguments:
+    ...    - ${x}: Relative x-coordinate to move the pointer to. It must be in
```
"Output-relative"? It's not clear what it's relative to.

hbatagelo replied:
Yes, changed to "output-relative".

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    - ${x}: Relative x-coordinate to move the pointer to. It must be in
+    ...      the range 0..1, where 0 represents the left edge, and 1 represents
+    ...      the right edge of the output.
+    ...    - ${y}: Relative y-coordinate to move the pointer to. It must be in
```
Same.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    Return:
+    ...    Absolute position of the pointer after the move, as a tuple (x, y)
+    ...    of integers.
+    ${position}=    WaylandHid.Move Pointer To Proportional    ${x}    ${y}
```
```suggestion
    ${position}=    Move_Pointer_To_Proportional    ${x}    ${y}
```

hbatagelo replied:
Changed to `HID.`

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    - ${delay} (optional): time to sleep after each step, in seconds.
+    ...      Default is 0.01.
+    [Arguments]    ${steps}=0    ${delay}=0.01
+    WaylandHid.Walk Pointer To Absolute    ${x}    ${y}    ${steps}    ${delay}
```
```suggestion
    Walk_Pointer_To_Absolute    ${x}    ${y}    ${steps}    ${delay}
```

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...      moved in increments of 16 pixels. Default is 0.
+    ...    - ${delay} (optional): time to sleep after each step, in seconds.
+    ...      Default is 0.01.
+    [Arguments]    ${steps}=0    ${delay}=0.01
```
I'd move `[Arguments]` up near the keyword, if possible?

hbatagelo replied:
Yes, done... although I'd still prefer the documentation to come first as a docstring. The `[Arguments]` tag in this case contains only the optional arguments.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    Return:
+    ...    Absolute position of the pointer after the walk, as a tuple (x, y)
+    ...    of integers.
+    [Arguments]    ${steps}=0    ${delay}=0.01
```
Here, too.

hbatagelo replied:
Done.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+        ${position}=    Set Variable    ${destination}
     END
-    RETURN    ${end_position}
+    WaylandHid.Walk Pointer To Absolute    ${position}[0]    ${position}[1]    ${steps}    ${delay}
```
```suggestion
    Walk_Pointer_To_Absolute    ${position}[0]    ${position}[1]    ${steps}    ${delay}
```

hbatagelo replied:
Changed to `HID.`

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    Return:
+    ...    Absolute position of the pointer after the walk, as a tuple (x, y)
+    ...    of integers.
+    [Arguments]    ${steps}=0    ${delay}=0.01
```
And here.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...    Absolute position of the pointer after the walk, as a tuple (x, y)
+    ...    of integers.
+    [Arguments]    ${steps}=0    ${delay}=0.01
+    ${position}=    WaylandHid.Walk Pointer To Proportional    ${x}    ${y}    ${steps}    ${delay}
```
```suggestion
    ${position}=    Walk_Pointer_To_Proportional    ${x}    ${y}    ${steps}    ${delay}
```

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...
+    ...    Arguments:
+    ...    - ${button}: Button to press (LEFT|RIGHT|MIDDLE).
+    WaylandHid.Press Pointer Button    ${button}
```
```suggestion
    Press_Pointer_Button    ${button}
```

hbatagelo replied:
Same.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+    ...
+    ...    Arguments:
+    ...    - ${button}: Button to release (LEFT|RIGHT|MIDDLE).
+    WaylandHid.Release Pointer Button    ${button}
```
```suggestion
    Release_Pointer_Button    ${button}
```

hbatagelo replied:
Same.

Saviq commented on mir-ci/mir_ci/robot_resources/screencopy.resource:
```diff
@@ -8,93 +9,161 @@ Library             ../robot_libraries/WaylandHid.py
+
+Release Buttons
+    [Documentation]    Release all buttons on the virtual pointer.
+    WaylandHid.Release Buttons
```
Will this work?

```suggestion
    Release_Buttons
```

hbatagelo replied:
Same.

Saviq reviewed (changes requested):
We need `ROBOT_AUTO_KEYWORDS=False`:

https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#creating-keywords

---

It's a shame that we have to repeat keywords in the resource just for the sake of documenting them, WDYT?

On the other hand, it kind-of serves as an interface?…

hbatagelo commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -19,44 +20,90 @@ class WaylandHid(VirtualPointer):
+
+        :param pixel_step: Step size in pixels to be used when a nonpositive
+                           number of steps is passed to walk_pointer_*.
+                           Default is 16.
```
I dropped `pixel_step`. We were using positive/nonpositive numbers to distinguish between number of steps and step size in pixels, but changing the semantics of a parameter based on the value doesn't seem right.

hbatagelo commented on mir-ci/mir_ci/robot_libraries/WaylandHid.py:
```diff
@@ -19,44 +20,90 @@ class WaylandHid(VirtualPointer):
+        return absolute_position
+
+    @keyword
+    async def walk_pointer_to_absolute(self, x: int, y: int, steps: int, delay: float) -> None:
```
I agree. I'll drop `steps` and use `step_distance` instead.

hbatagelo commented:
> We need `ROBOT_AUTO_KEYWORDS=False`:
> 
> https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#creating-keywords
> 
> It's a shame that we have to repeat keywords in the resource just for the sake of documenting them, WDYT?
> 
> On the other hand, it kind-of serves as an interface?…

We use the `@library` decorator which implicitly sets `ROBOT_AUTO_KEYWORDS` to `False`.

Well, I like the resource acting as a high-level interface so that `WaylandHid` doesn't need to depend on `Screencopy`.

Saviq reviewed (approved):
A'ight, let's get it in!
